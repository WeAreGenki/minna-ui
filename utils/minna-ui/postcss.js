'use strict'; // eslint-disable-line

module.exports = require('@minna-ui/postcss-config');
