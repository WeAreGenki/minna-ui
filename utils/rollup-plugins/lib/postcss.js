'use strict';

const merge = require('deepmerge');
const { dirname, join } = require('path');
const postcss = require('postcss');
const postcssrc = require('postcss-load-config');
const Purgecss = require('purgecss');
const { createFilter } = require('rollup-pluginutils');

/**
 * Rollup plugin to process any imported CSS via PostCSS and optionally remove
 * unused styles for significantly smaller CSS bundles.
 * @param {Object} opts User defined options.
 * @param {(Array<string>|Function)=} opts.content Page content.
 * @param {Object=} opts.context Base PostCSS options.
 * @param {Array<string>=} opts.exclude Files to exclude from CSS processing.
 * @param {Array<string>=} opts.include Files to include in CSS processing.
 * @param {Array<string>=} opts.content Files to parse for CSS classes.
 * @param {boolean=} opts.optimize Should output CSS be minified and cleaned?
 * @param {Array<string>=} opts.whitelist CSS classes to always keep.
 * @returns {Object} Rollup plugin
 */
function postcssRollup({
  content = [
    '__sapper__/build/*.html',
    '__sapper__/build/*.js',
    // FIXME: Using `dist` is the most reliable but requires 2 builds
    // 'dist/**/*.html',
    // 'dist/**/*.js',
    'src/**/*.html',
    'src/**/*.js',
  ],
  context = {},
  exclude = [],
  include = ['**/*.css'],
  optimize = process.env.NODE_ENV !== 'development',
  whitelist = [],
} = {}) {
  const filter = createFilter(include, exclude);

  return {
    name: 'postcss',

    async transform(source, id) {
      if (!filter(id)) return;

      try {
        const ctx = merge(
          { from: id, map: { annotation: false, inline: false }, to: id },
          context,
        );
        const { plugins, options } = await postcssrc(ctx);
        const result = await postcss(plugins).process(source, options);

        result.warnings().forEach((warn) => {
          this.warn(warn.toString(), { column: warn.column, line: warn.line });
        });

        // register sub-dependencies so rollup can monitor them for changes
        if (result.map) {
          const basePath = dirname(id);
          // TODO: Don't use PostCSS private API
          // eslint-disable-next-line no-underscore-dangle
          result.map._sources._array.forEach((dep) => {
            this.addWatchFile(join(basePath, dep));
          });
        }

        if (!optimize) {
          // eslint-disable-next-line consistent-return
          return {
            code: result.css,
            map: result.map,
          };
        }

        const purgecss = new Purgecss({
          content,
          css: [{ raw: result.css }],
          keyframes: true,
          whitelist,
        });

        const purged = purgecss.purge()[0];

        // eslint-disable-next-line consistent-return
        return {
          code: purged.css,
          map: purged.map,
        };
      } catch (err) {
        if (err.name === 'CssSyntaxError') {
          // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
          process.stderr.write(err.message + err.showSourceCode());
        } else {
          this.error(err);
        }
      }
    },
  };
}

module.exports = postcssRollup;
