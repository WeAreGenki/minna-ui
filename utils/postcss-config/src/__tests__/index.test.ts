/** @jest-environment node */

/* eslint-disable @typescript-eslint/no-non-null-assertion */

import fs from 'fs';
import path from 'path';
import { promisify } from 'util';
import postcss from 'postcss';
import postcssConfig from '../index';

// eslint-disable-next-line security/detect-non-literal-fs-filename
const readFile = promisify(fs.readFile);
const jestConfigPath = require.resolve('@minna-ui/jest-config');
const fixturesPath = path.join(path.dirname(jestConfigPath), 'fixtures');
const sourceCssPath = path.join(fixturesPath, 'import.css');
// const sourceCssMixinPath = path.join(fixturesPath, 'mixin.css');
// const mixinsPath = path.join(fixturesPath, 'css-mixins');

const options = {
  from: sourceCssPath,
  map: {
    // annotation: true,
    inline: false,
  },
};

let sourceCss = '';
// let sourceCssMixin = '';

beforeAll(async () => {
  // [sourceCss, sourceCssMixin] = await Promise.all([
  //   readFile(sourceCssPath, 'utf8'),
  //   readFile(sourceCssMixinPath, 'utf8'),
  // ]);
  sourceCss = await readFile(sourceCssPath, 'utf8');
});

describe('PostCSS config', () => {
  it('compiles valid CSS', async () => {
    expect.assertions(5);
    const output = postcss(postcssConfig()).process(sourceCss, options);
    await expect(output).resolves.toBeDefined();
    const result = await output;
    expect(result.processor!.plugins).not.toHaveLength(0);
    expect(result.opts!.from).toBeDefined();
    expect(result.map).toBeDefined();
    expect(result.css).toMatchSnapshot();
  });

  it('compiles CSS with optimize option true', async () => {
    expect.assertions(5);
    const output = postcss(postcssConfig({ optimize: true })).process(sourceCss, options);
    await expect(output).resolves.toBeDefined();
    const result = await output;
    expect(result.processor!.plugins).not.toHaveLength(0);
    expect(result.opts!.from).toBeDefined();
    expect(result.map).toBeDefined();
    expect(result.css).toMatchSnapshot();
  });

  it('compiles CSS with optimize option false', async () => {
    expect.assertions(5);
    const output = postcss(postcssConfig({ optimize: false })).process(sourceCss, options);
    await expect(output).resolves.toBeDefined();
    const result = await output;
    expect(result.processor!.plugins).not.toHaveLength(0);
    expect(result.opts!.from).toBeDefined();
    expect(result.map).toBeDefined();
    expect(result.css).toMatchSnapshot();
  });

  it('compiles CSS with unsafe option true', async () => {
    expect.assertions(5);
    const output = postcss(postcssConfig({ optimize: true, unsafe: true })).process(
      sourceCss,
      options,
    );
    await expect(output).resolves.toBeDefined();
    const result = await output;
    expect(result.processor!.plugins).not.toHaveLength(0);
    expect(result.opts!.from).toBeDefined();
    expect(result.map).toBeDefined();
    expect(result.css).toMatchSnapshot();
  });

  it('compiles CSS with unsafe option false', async () => {
    expect.assertions(5);
    const output = postcss(postcssConfig({ optimize: true, unsafe: false })).process(
      sourceCss,
      options,
    );
    await expect(output).resolves.toBeDefined();
    const result = await output;
    expect(result.processor!.plugins).not.toHaveLength(0);
    expect(result.opts!.from).toBeDefined();
    expect(result.map).toBeDefined();
    expect(result.css).toMatchSnapshot();
  });

  it('compiles CSS with debug option true', async () => {
    expect.assertions(6);
    const spy = jest.spyOn(global.console, 'log');
    spy.mockImplementation(() => {});
    const output = postcss(postcssConfig({ debug: true })).process(sourceCss, options);
    await expect(output).resolves.toBeDefined();
    const result = await output;
    expect(result.processor!.plugins).not.toHaveLength(0);
    expect(result.opts!.from).toBeDefined();
    expect(result.map).toBeDefined();
    expect(result.css).toMatchSnapshot();
    expect(spy).toHaveBeenCalledTimes(2); // should log import path twice
    spy.mockRestore();
  });

  it('compiles CSS with debug option false', async () => {
    expect.assertions(6);
    const spy = jest.spyOn(global.console, 'log');
    spy.mockImplementation(() => {});
    const output = postcss(postcssConfig({ debug: false })).process(sourceCss, options);
    await expect(output).resolves.toBeDefined();
    const result = await output;
    expect(result.processor!.plugins).not.toHaveLength(0);
    expect(result.opts!.from).toBeDefined();
    expect(result.map).toBeDefined();
    expect(result.css).toMatchSnapshot();
    expect(spy).not.toHaveBeenCalled();
    spy.mockRestore();
  });

  it.todo('writes correct import path');
  it.todo('writes correct import path with importAlias option');
  it.todo('compiles with importPaths option');
  it.todo('compiles with custom options');
  it.todo('supports $var compile-time variables');
  it.todo('supports --var runtime variables');
  it.todo('supports @import');
  it.todo('supports @mixin');
  it.todo('supports inline comments');
  it.todo('supports inline comments in imports');
});
