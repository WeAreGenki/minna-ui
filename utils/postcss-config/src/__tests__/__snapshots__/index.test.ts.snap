// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostCSS config compiles CSS with debug option false 1`] = `
"/* red */
body {
  margin: 0;
}
/* stylelint-disable selector-max-id */
#target {
  color: rgb(255, 0, 0);
}
#target .focus,
  #target:focus {
    background: rgb(0, 0, 0);
  }
.wrapper + #target {
    background: rgb(33, 33, 33);
  }
/* stylelint-enable */

/*# sourceMappingURL=utils/jest-config/__fixtures__/import.css.map */"
`;

exports[`PostCSS config compiles CSS with debug option true 1`] = `
"/* red */
body {
  margin: 0;
}
/* stylelint-disable selector-max-id */
#target {
  color: rgb(255, 0, 0);
}
#target .focus,
  #target:focus {
    background: rgb(0, 0, 0);
  }
.wrapper + #target {
    background: rgb(33, 33, 33);
  }
/* stylelint-enable */

/*# sourceMappingURL=utils/jest-config/__fixtures__/import.css.map */"
`;

exports[`PostCSS config compiles CSS with optimize option false 1`] = `
"/* red */
body {
  margin: 0;
}
/* stylelint-disable selector-max-id */
#target {
  color: rgb(255, 0, 0);
}
#target .focus,
  #target:focus {
    background: rgb(0, 0, 0);
  }
.wrapper + #target {
    background: rgb(33, 33, 33);
  }
/* stylelint-enable */

/*# sourceMappingURL=utils/jest-config/__fixtures__/import.css.map */"
`;

exports[`PostCSS config compiles CSS with optimize option true 1`] = `
"body{margin:0}#target{color:red}#target .focus,#target:focus{background:#000}.wrapper+#target{background:#212121}
/*# sourceMappingURL=utils/jest-config/__fixtures__/import.css.map */"
`;

exports[`PostCSS config compiles CSS with unsafe option false 1`] = `
"body{margin:0}#target{color:red}#target .focus,#target:focus{background:#000}.wrapper+#target{background:#212121}
/*# sourceMappingURL=utils/jest-config/__fixtures__/import.css.map */"
`;

exports[`PostCSS config compiles CSS with unsafe option true 1`] = `
"body{margin:0}#target{color:red}#target .focus,#target:focus{background:#000}.wrapper+#target{background:#212121}
/*# sourceMappingURL=utils/jest-config/__fixtures__/import.css.map */"
`;

exports[`PostCSS config compiles valid CSS 1`] = `
"/* red */
body {
  margin: 0;
}
/* stylelint-disable selector-max-id */
#target {
  color: rgb(255, 0, 0);
}
#target .focus,
  #target:focus {
    background: rgb(0, 0, 0);
  }
.wrapper + #target {
    background: rgb(33, 33, 33);
  }
/* stylelint-enable */

/*# sourceMappingURL=utils/jest-config/__fixtures__/import.css.map */"
`;
